packages/@react-aria/landmark/src/useLandmark.ts

```diff
@@ -0,0 +1,261 @@
+/*
+ * Copyright 2022 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import {AriaLabelingProps} from '@react-types/shared';
+import {HTMLAttributes, MutableRefObject, useEffect} from 'react';
+import {useLayoutEffect} from '@react-aria/utils';
+
+export type AriaLandmarkRole = 'main' | 'region' | 'search' | 'navigation' | 'form' | 'banner' | 'contentinfo' | 'complementary';
+
+export interface AriaLandmarkProps extends AriaLabelingProps {
+  role: AriaLandmarkRole
+}
+
+interface LandmarkAria {
+  landmarkProps: HTMLAttributes<HTMLElement>
+}
+
+type Landmark = {
+  ref: MutableRefObject<HTMLElement>,
+  role: AriaLandmarkRole,
+  label?: string,
+  lastFocused?: HTMLElement
+};
+
+class LandmarkManager {
+  private landmarks: Array<Landmark> = [];
+  private static instance: LandmarkManager;
+
+  private constructor() {}
+
+  public static getInstance(): LandmarkManager {
+    if (!LandmarkManager.instance) {
+      LandmarkManager.instance = new LandmarkManager();
+      LandmarkManager.instance.setup();
+    }
+
+    return LandmarkManager.instance;
+  }
+
+  private setup() {
+    document.addEventListener('keydown', LandmarkManager.getInstance().f6Handler.bind(LandmarkManager.getInstance()), {capture: true});
+    document.addEventListener('focusin', LandmarkManager.getInstance().focusinHandler.bind(LandmarkManager.getInstance()), {capture: true});
+  }
+
+  private onBlurLandmark(event) {
+    let {target: landmark} = event;
+    landmark.removeAttribute('tabindex');
+    landmark.removeEventListener('blur', this.onBlurLandmark);
+  }
+
+  private focusLandmark(landmark: HTMLElement) {
+    landmark.setAttribute('tabindex', '-1');
+    landmark.addEventListener('blur', this.onBlurLandmark);
+    landmark.focus();
+  }
```

Comment:
If a user supplies a tabIndex already, then we probably want to restore the last tabIndex before we change it.

I have an idea to give a bit more control to users and let React manager things a bit more. Let me know what you think. If you want, you can try it in a new branch based on this PR. We'll need to run it by @devongovett  as well, I don't think we do custom events anywhere else and I want to make sure we introduce them for a good reason if we're going to do that.

We fire a custom event for focusLandmark, 'landmarkFocus'

then, down in useLandmark, we setup a listener for this event. When one occurs, it sets the state (isLandmarkFocused for example) of that landmark to true. That will cause a re-render and in the returned object from the hook we would have something like
```
return {
  role,
  tabIndex: isLandmarkFocused ? -1 : undefined
}
```

We'd also need a useEffect that runs when `isLandmarkFocused` changes that would call `ref.current.focus()` if `isLandmarkFocused` is true

useLandmark would also setup an onblur listener, which would just set `isLandmarkFocused` to false

Reply:
https://developer.mozilla.org/en-US/docs/Web/Events/Creating_and_triggering_events

Reply:
I like this idea a lot. I'll give it a try and we can go in that direction if there aren't any holdups.

Reply:
@snowystinger Give this a go at https://github.com/adobe/react-spectrum/pull/2878, let me know what you think.